File: dom/nfc/MozNdefRecord.cpp
/* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
/* vim:set ts=2 sw=2 sts=2 et cindent: */
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* Copyright © 2013 Deutsche Telekom, Inc. */

#include "MozNdefRecord.h"
#include "mozilla/dom/MozNdefRecordBinding.h"
#include "mozilla/HoldDropJSObjects.h"
#include "nsContentUtils.h"


namespace mozilla {
namespace dom {

NS_IMPL_CYCLE_COLLECTION_WRAPPERCACHE_1(MozNdefRecord, mWindow)
NS_IMPL_CYCLE_COLLECTING_ADDREF(MozNdefRecord)
NS_IMPL_CYCLE_COLLECTING_RELEASE(MozNdefRecord)
NS_INTERFACE_MAP_BEGIN_CYCLE_COLLECTION(MozNdefRecord)
  NS_WRAPPERCACHE_INTERFACE_MAP_ENTRY
  NS_INTERFACE_MAP_ENTRY(nsISupports)
NS_INTERFACE_MAP_END

void
MozNdefRecord::HoldData()
{
  mozilla::HoldJSObjects(this);
}

void
MozNdefRecord::DropData()
{
  if (mType) {
    mType = nullptr;
  }
  if (mId) {
    mId = nullptr;
  }
  if (mPayload) {
    mPayload = nullptr;
  }
  mozilla::DropJSObjects(this);
}

/* static */
already_AddRefed<MozNdefRecord>
MozNdefRecord::Constructor(const GlobalObject& aGlobal,
  uint8_t aTnf, const Uint8Array& aType, const Uint8Array& aId, const Uint8Array& aPayload,
  ErrorResult& aRv)
{
  nsCOMPtr<nsPIDOMWindow> win = do_QueryInterface(aGlobal.GetAsSupports());
  if (!win) {
    aRv.Throw(NS_ERROR_FAILURE);
    return nullptr;
  }

  nsRefPtr<MozNdefRecord> ndefrecord = new MozNdefRecord(aGlobal.GetContext(), win, aTnf, aType, aId, aPayload);
  if (!ndefrecord) {
    aRv.Throw(NS_ERROR_FAILURE);
    return nullptr;
  }
  return ndefrecord.forget();
}

MozNdefRecord::MozNdefRecord(JSContext* aCx, nsPIDOMWindow* aWindow, uint8_t aTnf, const Uint8Array& aType, const Uint8Array& aId, const Uint8Array& aPayload)
  : mTnf(aTnf)
{
  mWindow = aWindow; // For GetParentObject()

  mType = Uint8Array::Create(aCx, this, aType.Length(), aType.Data());
  mId = Uint8Array::Create(aCx, this, aId.Length(), aId.Data());
  mPayload = Uint8Array::Create(aCx, this, aPayload.Length(), aPayload.Data());

  SetIsDOMBinding();
  HoldData();
}

MozNdefRecord::~MozNdefRecord()
{
  DropData();
}

JSObject*
MozNdefRecord::WrapObject(JSContext* aCx, JS::Handle<JSObject*> aScope)
{
  return MozNdefRecordBinding::Wrap(aCx, aScope, this);
}

} // namespace dom
} // namespace mozilla

File: dom/nfc/MozNdefRecord.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright © 2013, Deutsche Telekom, Inc.
 */

"use strict";

const DEBUG = false;
function debug(s) {
  if (DEBUG) dump("-*- Nfc MozNdefRecord: " + s + "\n");
}

const Cc = Components.classes;
const Ci = Components.interfaces;
const Cu = Components.utils;

Cu.import("resource://gre/modules/XPCOMUtils.jsm");
Cu.import("resource://gre/modules/Services.jsm");

function MozNdefRecord() {
  debug("XXXX In MozNdefRecord Constructor");
  this._nfcContentHelper = Cc["@mozilla.org/nfc/content-helper;1"]
                             .getService(Ci.nsINfcContentHelper);
}
MozNdefRecord.prototype = {
  _tnf: null,
  _type: null,
  _id: null,
  _payload: null,

  init: function init(aWindow) {
  },

  __init: function(tnf, type, id, payload) {
    this._tnf = tnf;
    this._type = type;
    this._id = id;
    this._payload = payload;
  },

  get tnf() {
    return this._tnf;
  },
  
  set tnf(v) {
    this._tnf = v;
  },
  
  get type() {
    return this._type;
  },
  
  set type(v) {
    this._type = v;
  },
  
  get id() {
    return this._id;
  },
  
  set id(v) {
    this._id = v;
  },
  
  get payload() {
    return this._payload;
  },
  
  set payload(v) {
    this._payload = v;
  },

  classID: Components.ID("{4E7DD439-D639-4DA9-8F8E-D019CAD9EAED}"),
  contractID: "@mozilla.org/nfc/NdefRecord;1",
  QueryInterface: XPCOMUtils.generateQI([Ci.nsISupports,
                                         Ci.nsIDOMGlobalPropertyInitializer]),
};

this.NSGetFactory = XPCOMUtils.generateNSGetFactory([MozNdefRecord]);
